feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,364 @@
+			LbPolicy:             v2.Cluster_CLUSTER_PROVIDED,
+		},
+	}
+	filterChains := []*listener.FilterChain{}
+	virtualHosts := []*v2route.VirtualHost{getVirtualHost(virtualHostName, []string{"*"}, []*core.HeaderValueOption{})}
+
+	pbst, err := ptypes.MarshalAny(getHttpConnectionManager(defaultRouteName, virtualHosts))
+	if err != nil {
+		glog.Errorf("Couldn't marshal HTTP connection manager")
+		return
+	}
+	filterChains = append(filterChains, &listener.FilterChain{
+		FilterChainMatch: &listener.FilterChainMatch{},
+		Filters: []*listener.Filter{{
+			Name: wellknown.HTTPConnectionManager,
+			ConfigType: &listener.Filter_TypedConfig{
+				TypedConfig: pbst,
+			},
+		}},
+	})
+
+	for _, req := range ues {
+		var ue_ip_addr = string(req.UeIp.Address)
+		requestHeadersToAdd := []*core.HeaderValueOption{}
+
+		glog.Infof("Adding UE - " + ue_ip_addr)
+
+		for _, header := range req.Headers {
+			requestHeadersToAdd = append(requestHeadersToAdd, &core.HeaderValueOption{
+				Header: &core.HeaderValue{
+					Key:   header.Name,
+					Value: header.Value,
+				},
+			})
+		}
+
+		virtualHosts = []*v2route.VirtualHost{getVirtualHost(virtualHostName, req.Websites, requestHeadersToAdd)}
+
+		pbst, err := ptypes.MarshalAny(getHttpConnectionManager(routeConfigName, virtualHosts))
+		if err != nil {
+			glog.Errorf("Couldn't marshal UE HTTP connection manager")
+			continue
+		}
+
+		filterChains = append(filterChains, &listener.FilterChain{
+			FilterChainMatch: &listener.FilterChainMatch{
+				SourcePrefixRanges: []*core.CidrRange{{
+					AddressPrefix: ue_ip_addr,
+					PrefixLen:     &wrappers.UInt32Value{Value: 32},
+				}},
+			},
+			Filters: []*listener.Filter{{
+				Name: wellknown.HTTPConnectionManager,
+				ConfigType: &listener.Filter_TypedConfig{
+					TypedConfig: pbst,
+				},
+			}},
+		})
+
+	}
+
+	o_src := &orig_src.OriginalSrc{}
+	mo_src, err := ptypes.MarshalAny(o_src)
+	if err != nil {
+		glog.Errorf("Couldn't marshal OriginalSrc")
+		return
+	}
+
+	glog.Infof("Creating listener " + listenerName)
+	var listener = []cache.Resource{
+		&v2.Listener{
+			Name:        listenerName,
+			Transparent: &wrappers.BoolValue{Value: true},
+			Address: &core.Address{
+				Address: &core.Address_SocketAddress{
+					SocketAddress: &core.SocketAddress{
+						Address: any_addr,
+						PortSpecifier: &core.SocketAddress_PortValue{
+							PortValue: httpPort,
+						},
+					},
+				},
+			},
+			FilterChains: filterChains,
+			ListenerFilters: []*listener.ListenerFilter{
+				{
+					Name: "envoy.filters.listener.original_dst",
+				},
+				{
+					Name: "envoy.filters.listener.original_src",
+					ConfigType: &listener.ListenerFilter_TypedConfig{
+						TypedConfig: mo_src,
+					},
+				},
+			},
+		}}
+
+	// Save snapshot
+	atomic.AddInt32(&cli.version, 1)
+	glog.Infof("Saved snapshot version " + fmt.Sprint(cli.version))
+	snap := cache.NewSnapshot(fmt.Sprint(cli.version), nil, cluster, nil, listener, nil)
```

Comment:
I think cluster should also be define right before this line if possible. 
This way we dont have to go up and down to see what is cluster